File: backend/src/services/pricing.js
const TAX_RATE = parseFloat(process.env.TAX_RATE || '0.08');
const FREE_SHIPPING_THRESHOLD = parseFloat(process.env.FREE_SHIPPING_THRESHOLD || '100');
const FLAT_SHIPPING_FEE = parseFloat(process.env.FLAT_SHIPPING_FEE || '10');

function calculateItemPrice(basePrice, discountPrice) {
  const flatDiscount = parseFloat(discountPrice || 0);
  let finalPrice = basePrice - flatDiscount;
  if (finalPrice < 0) finalPrice = 0;
  return finalPrice;
}

function calculateOrderTotals(subtotal, discountAmount) {
  const taxableSubtotal = subtotal - discountAmount;
  const taxAmount = taxableSubtotal * TAX_RATE;
  const shippingAmount = taxableSubtotal >= FREE_SHIPPING_THRESHOLD ? 0.0 : FLAT_SHIPPING_FEE;
  const finalTotal = taxableSubtotal + taxAmount + shippingAmount;

  return {
    taxableSubtotal,
    taxAmount,
    shippingAmount,
    finalTotal,
  };
}

module.exports = {
  TAX_RATE,
  FREE_SHIPPING_THRESHOLD,
  FLAT_SHIPPING_FEE,
  calculateItemPrice,
  calculateOrderTotals,
};

File: backend-fastapi/services/pricing.py
"""
Pricing service — tax, shipping, and discount calculations.
"""
import os
from decimal import Decimal


def get_tax_rate() -> Decimal:
    return Decimal(os.getenv("TAX_RATE", "0.08"))


def get_free_shipping_threshold() -> Decimal:
    return Decimal(os.getenv("FREE_SHIPPING_THRESHOLD", "100"))


def get_flat_shipping_fee() -> Decimal:
    return Decimal(os.getenv("FLAT_SHIPPING_FEE", "10"))


def calculate_item_price(base_price: Decimal, discount_price: Decimal) -> Decimal:
    """Calculate the final unit price after applying the flat discount."""
    discount = Decimal(str(discount_price)) if discount_price else Decimal("0")
    final = Decimal(str(base_price)) - discount
    return max(final, Decimal("0"))


def calculate_order_totals(subtotal: Decimal, discount_amount: Decimal) -> dict:
    """Calculate tax, shipping, and final total."""
    after_discount = subtotal - discount_amount
    if after_discount < 0:
        after_discount = Decimal("0")

    tax_rate = get_tax_rate()
    tax_amount = (after_discount * tax_rate).quantize(Decimal("0.01"))

    threshold = get_free_shipping_threshold()
    flat_fee = get_flat_shipping_fee()
    shipping_amount = Decimal("0") if after_discount >= threshold else flat_fee

    final_total = after_discount + tax_amount + shipping_amount

    return {
        "tax_amount": tax_amount,
        "shipping_amount": shipping_amount,
        "final_total": final_total.quantize(Decimal("0.01")),
    }
